### ISSUE TITLE
[Cache] PhpFilesAdapter fails to load the cache

### ISSUE THREAD

**Issue by ghost** [symfony/symfony#45244]
### Symfony version(s) affected

6.0

### Description

Hi,

The problem is quite simple, the [PhpFilesAdapter](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Cache/Adapter/PhpFilesAdapter.php) fails to load the cache.

I think these issues are related to mine:
- [#1433](https://github.com/doctrine/DoctrineBundle/issues/1433)
- #44256

**[!!!] VERY IMPORTANT [!!!]** 
To be able to see the error message of the issue, you have to comment [this line](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Cache/Adapter/PhpFilesAdapter.php#L132) because the error handler provided by the Symfony component doesn't work (no idea why). Just like this: 
```php
        if (!$missingIds) {
            return $values;
        }
        // /!\ Comment the line below to see the error /!\
        // set_error_handler($this->includeHandler);
        try {
            $getExpiry = true;

            foreach ($missingIds as $k => $id) {
                try {
                    $file = $this->files[$id] ??= $this->getFile($id);

                    if (isset(self::$valuesCache[$file])) {
                        [$expiresAt, $this->values[$id]] = self::$valuesCache[$file];
                    } elseif (\is_array($expiresAt = include $file)) {
                        if ($this->appendOnly) {
                            self::$valuesCache[$file] = $expiresAt;
                        }

                        [$expiresAt, $this->values[$id]] = $expiresAt;
                    } elseif ($now < $expiresAt) {
                        $this->values[$id] = new LazyValue($file);
                    }

                    if ($now >= $expiresAt) {
                        unset($this->values[$id], $missingIds[$k], self::$valuesCache[$file]);
                    }
                } catch (\ErrorException $e) {
                    unset($missingIds[$k]);
                }
            }
        } finally {
            restore_error_handler();
        }
```

Here is the error thrown by the adapter:
```
PHP Warning:  include(C:\symfony_cache_issue\cache\bar\V\K\qy98RaUfrjUdrbEirLuw): Failed to open stream: No such file or directory in C:\symfony_cache_issue\vendor\symfony\cache\Adapter\PhpFilesAdapter.php 
on line 173
PHP Stack trace:
PHP   1. {main}() C:\symfony_cache_issue\index.php:0
PHP   2. Symfony\Component\Cache\Adapter\AbstractAdapter->get($key = 'foo', $callback = class Closure { virtual $closure = "{closure}",  }, $beta = *uninitialized*, $metadata = *uninitialized*) C:\symfony_cache_issue\index.php:9
```

Thank you in advance for your assistance.

Regards!

### How to reproduce

You just need this short code to reproduce the issue.
```php
<?php

use Symfony\Component\Cache\Adapter\PhpFilesAdapter;

require __DIR__ . '/vendor/autoload.php';

$pool = new PhpFilesAdapter('bar', 0, __DIR__ . '/cache');

$data = $pool->get('foo',  fn (): mixed => [1, 2, 3]);

var_dump($data);
```

### Possible Solution

I have no clue what the problem is.

### Additional Context

PHP Version:
```
PHP 8.1.2 (cli) (built: Jan 19 2022 10:18:23) (ZTS Visual C++ 2019 x64)
Copyright (c) The PHP Group
Zend Engine v4.1.2, Copyright (c) Zend Technologies
    with Xdebug v3.1.2, Copyright (c) 2002-2021, by Derick Rethans     
```

Content of the `composer.json` file:
```
{
    "require": {
        "symfony/cache": "^6.0"
    }
}
```

Screenshot of the code:
![symfony_cache_issue](https://user-images.githubusercontent.com/59090546/151701586-e13c3286-7cb1-4cfb-bf73-8d8f1449c38a.png)

**Comment by nicolas-grekas** [comment]
I totally fail to understand this issue. What you're describing is the expected behavior of the code. This line serves a purpose. Commenting it doesn't highlight any issue. Commenting it creates the issue actually...

On first run, the cache is empty, so there is nothing to load and that's it. The line that you comment is here precisely to silence the warning, since that very behavior is expected when running with an empty cache.

Then, when calling the script a second time, the item has been persisted and the value is loaded from the cache.

**Comment by ghost** [comment]
> I totally fail to understand this issue. What you're describing is the expected behavior of the code. This line serves a purpose. Commenting it doesn't highlight any issue. Commenting it creates the issue actually...
> 
> On first run, the cache is empty, so there is nothing to load and that's it. The line that you comment is here precisely to silence the warning, since that very behavior is expected when running with an empty cache.
> 
> Then, when calling the script a second time, the item has been persisted and the value is loaded from the cache.

@nicolas-grekas 

Hi, thanks for you answer.

I commented on this line because nothing happens without it, and I tried to identify the issue I have.

Look at my screenshots below. I have removed the comment from the line in question, again, to show you the extent of my issue. Why does my code stop with adapter `PhpFilesAdapter` (the `var_dump` does not display my data even after 3 executions of the script), but not with adapter `FilesystemAdapter` which works correctly (the `var_dump` show my data after the first execution) ? Aren't adapters supposed to work the same way? Did I miss something? Why is nothing happening with the adapter `PhpFilesAdapter`?

`PhpFilesAdapter` behavior (nothing happens and my script shut down, even after 3 script executions)
![1](https://user-images.githubusercontent.com/59090546/151781256-9e511dde-745d-4825-9b5d-47efdace2058.png)

`FilesystemAdapter` behavior (all is ok, i get my data after the first script execution)
![2](https://user-images.githubusercontent.com/59090546/151781257-4bf75459-e258-457a-96c3-a880bef54b97.png)

Thanks for your assistance.

**Comment by nicolas-grekas** [comment]
This is clearly not expected, but I don't reproduce the behavior you have, so this might be related to your version of PHP, or to your opcache php.ini setting? Sorry I cannot help if I don't reproduce. Does anyone else see this behavior?

**Comment by ghost** [comment]
> This is clearly not expected, but I don't reproduce the behavior you have, so this might be related to your version of PHP, or to your opcache php.ini setting? Sorry I cannot help if I don't reproduce. Does anyone else see this behavior?

@nicolas-grekas 

My **issue** occurs with **PHP 8.1.2**, while it **works** perfectly with **PHP 8.0.13**. See the screenshots below.

PHP 8.0.13:
![1](https://user-images.githubusercontent.com/59090546/151783878-328fb00f-83ee-476c-b78c-fd9d8860f278.png)

```ini
;zend_extension=opcache

[opcache]
; Determines if Zend OPCache is enabled
;opcache.enable=1

; Determines if Zend OPCache is enabled for the CLI version of PHP
;opcache.enable_cli=0

; The OPcache shared memory storage size.
;opcache.memory_consumption=128

; The amount of memory for interned strings in Mbytes.
;opcache.interned_strings_buffer=8

; The maximum number of keys (scripts) in the OPcache hash table.
; Only numbers between 200 and 1000000 are allowed.
;opcache.max_accelerated_files=10000

; The maximum percentage of "wasted" memory until a restart is scheduled.
;opcache.max_wasted_percentage=5

; When this directive is enabled, the OPcache appends the current working
; directory to the script key, thus eliminating possible collisions between
; files with the same name (basename). Disabling the directive improves
; performance, but may break existing applications.
;opcache.use_cwd=1

; When disabled, you must reset the OPcache manually or restart the
; webserver for changes to the filesystem to take effect.
;opcache.validate_timestamps=1

; How often (in seconds) to check file timestamps for changes to the shared
; memory storage allocation. ("1" means validate once per second, but only
; once per request. "0" means always validate)
;opcache.revalidate_freq=2

; Enables or disables file search in include_path optimization
;opcache.revalidate_path=0

; If disabled, all PHPDoc comments are dropped from the code to reduce the
; size of the optimized code.
;opcache.save_comments=1

; If enabled, compilation warnings (including notices and deprecations) will
; be recorded and replayed each time a file is included. Otherwise, compilation
; warnings will only be emitted when the file is first cached.
;opcache.record_warnings=0

; Allow file existence override (file_exists, etc.) performance feature.
;opcache.enable_file_override=0

; A bitmask, where each bit enables or disables the appropriate OPcache
; passes
;opcache.optimization_level=0x7FFFBFFF

;opcache.dups_fix=0

; The location of the OPcache blacklist file (wildcards allowed).
; Each OPcache blacklist file is a text file that holds the names of files
; that should not be accelerated. The file format is to add each filename
; to a new line. The filename may be a full path or just a file prefix
; (i.e., /var/www/x  blacklists all the files and directories in /var/www
; that start with 'x'). Line starting with a ; are ignored (comments).
;opcache.blacklist_filename=

; Allows exclusion of large files from being cached. By default all files
; are cached.
;opcache.max_file_size=0

; Check the cache checksum each N requests.
; The default value of "0" means that the checks are disabled.
;opcache.consistency_checks=0

; How long to wait (in seconds) for a scheduled restart to begin if the cache
; is not being accessed.
;opcache.force_restart_timeout=180

; OPcache error_log file name. Empty string assumes "stderr".
;opcache.error_log=

; All OPcache errors go to the Web server log.
; By default, only fatal errors (level 0) or errors (level 1) are logged.
; You can also enable warnings (level 2), info messages (level 3) or
; debug messages (level 4).
;opcache.log_verbosity_level=1

; Preferred Shared Memory back-end. Leave empty and let the system decide.
;opcache.preferred_memory_model=

; Protect the shared memory from unexpected writing during script execution.
; Useful for internal debugging only.
;opcache.protect_memory=0

; Allows calling OPcache API functions only from PHP scripts which path is
; started from specified string. The default "" means no restriction
;opcache.restrict_api=

; Mapping base of shared memory segments (for Windows only). All the PHP
; processes have to map shared memory into the same address space. This
; directive allows to manually fix the "Unable to reattach to base address"
; errors.
;opcache.mmap_base=

; Facilitates multiple OPcache instances per user (for Windows only). All PHP
; processes with the same cache ID and user share an OPcache instance.
;opcache.cache_id=

; Enables and sets the second level cache directory.
; It should improve performance when SHM memory is full, at server restart or
; SHM reset. The default "" disables file based caching.
;opcache.file_cache=

; Enables or disables opcode caching in shared memory.
;opcache.file_cache_only=0

; Enables or disables checksum validation when script loaded from file cache.
;opcache.file_cache_consistency_checks=1

; Implies opcache.file_cache_only=1 for a certain process that failed to
; reattach to the shared memory (for Windows only). Explicitly enabled file
; cache is required.
;opcache.file_cache_fallback=1

; Enables or disables copying of PHP code (text segment) into HUGE PAGES.
; This should improve performance, but requires appropriate OS configuration.
;opcache.huge_code_pages=0

; Validate cached file permissions.
;opcache.validate_permission=0

; Prevent name collisions in chroot'ed environment.
;opcache.validate_root=0

; If specified, it produces opcode dumps for debugging different stages of
; optimizations.
;opcache.opt_debug_level=0

; Specifies a PHP script that is going to be compiled and executed at server
; start-up.
; http://php.net/opcache.preload
;opcache.preload=

; Preloading code as root is not allowed for security reasons. This directive
; facilitates to let the preloading to be run as another user.
; http://php.net/opcache.preload_user
;opcache.preload_user=

; Prevents caching files that are less than this number of seconds old. It
; protects from caching of incompletely updated files. In case all file updates
; on your site are atomic, you may increase performance by setting it to "0".
;opcache.file_update_protection=2

; Absolute path used to store shared lockfiles (for *nix only).
;opcache.lockfile_path=/tmp
```

PHP 8.1.2:
![2](https://user-images.githubusercontent.com/59090546/151783901-54fa1a73-a1a4-4fe5-90c6-bcfe0a2710e3.png)

```ini
;zend_extension=opcache
[opcache]
; Determines if Zend OPCache is enabled
;opcache.enable=1

; Determines if Zend OPCache is enabled for the CLI version of PHP
;opcache.enable_cli=0

; The OPcache shared memory storage size.
;opcache.memory_consumption=128

; The amount of memory for interned strings in Mbytes.
;opcache.interned_strings_buffer=8

; The maximum number of keys (scripts) in the OPcache hash table.
; Only numbers between 200 and 1000000 are allowed.
;opcache.max_accelerated_files=10000

; The maximum percentage of "wasted" memory until a restart is scheduled.
;opcache.max_wasted_percentage=5

; When this directive is enabled, the OPcache appends the current working
; directory to the script key, thus eliminating possible collisions between
; files with the same name (basename). Disabling the directive improves
; performance, but may break existing applications.
;opcache.use_cwd=1

; When disabled, you must reset the OPcache manually or restart the
; webserver for changes to the filesystem to take effect.
;opcache.validate_timestamps=1

; How often (in seconds) to check file timestamps for changes to the shared
; memory storage allocation. ("1" means validate once per second, but only
; once per request. "0" means always validate)
;opcache.revalidate_freq=2

; Enables or disables file search in include_path optimization
;opcache.revalidate_path=0

; If disabled, all PHPDoc comments are dropped from the code to reduce the
; size of the optimized code.
;opcache.save_comments=1

; If enabled, compilation warnings (including notices and deprecations) will
; be recorded and replayed each time a file is included. Otherwise, compilation
; warnings will only be emitted when the file is first cached.
;opcache.record_warnings=0

; Allow file existence override (file_exists, etc.) performance feature.
;opcache.enable_file_override=0

; A bitmask, where each bit enables or disables the appropriate OPcache
; passes
;opcache.optimization_level=0x7FFFBFFF

;opcache.dups_fix=0

; The location of the OPcache blacklist file (wildcards allowed).
; Each OPcache blacklist file is a text file that holds the names of files
; that should not be accelerated. The file format is to add each filename
; to a new line. The filename may be a full path or just a file prefix
; (i.e., /var/www/x  blacklists all the files and directories in /var/www
; that start with 'x'). Line starting with a ; are ignored (comments).
;opcache.blacklist_filename=

; Allows exclusion of large files from being cached. By default all files
; are cached.
;opcache.max_file_size=0

; Check the cache checksum each N requests.
; The default value of "0" means that the checks are disabled.
;opcache.consistency_checks=0

; How long to wait (in seconds) for a scheduled restart to begin if the cache
; is not being accessed.
;opcache.force_restart_timeout=180

; OPcache error_log file name. Empty string assumes "stderr".
;opcache.error_log=

; All OPcache errors go to the Web server log.
; By default, only fatal errors (level 0) or errors (level 1) are logged.
; You can also enable warnings (level 2), info messages (level 3) or
; debug messages (level 4).
;opcache.log_verbosity_level=1

; Preferred Shared Memory back-end. Leave empty and let the system decide.
;opcache.preferred_memory_model=

; Protect the shared memory from unexpected writing during script execution.
; Useful for internal debugging only.
;opcache.protect_memory=0

; Allows calling OPcache API functions only from PHP scripts which path is
; started from specified string. The default "" means no restriction
;opcache.restrict_api=

; Mapping base of shared memory segments (for Windows only). All the PHP
; processes have to map shared memory into the same address space. This
; directive allows to manually fix the "Unable to reattach to base address"
; errors.
;opcache.mmap_base=

; Facilitates multiple OPcache instances per user (for Windows only). All PHP
; processes with the same cache ID and user share an OPcache instance.
;opcache.cache_id=

; Enables and sets the second level cache directory.
; It should improve performance when SHM memory is full, at server restart or
; SHM reset. The default "" disables file based caching.
;opcache.file_cache=

; Enables or disables opcode caching in shared memory.
;opcache.file_cache_only=0

; Enables or disables checksum validation when script loaded from file cache.
;opcache.file_cache_consistency_checks=1

; Implies opcache.file_cache_only=1 for a certain process that failed to
; reattach to the shared memory (for Windows only). Explicitly enabled file
; cache is required.
;opcache.file_cache_fallback=1

; Enables or disables copying of PHP code (text segment) into HUGE PAGES.
; This should improve performance, but requires appropriate OS configuration.
;opcache.huge_code_pages=0

; Validate cached file permissions.
;opcache.validate_permission=0

; Prevent name collisions in chroot'ed environment.
;opcache.validate_root=0

; If specified, it produces opcode dumps for debugging different stages of
; optimizations.
;opcache.opt_debug_level=0

; Specifies a PHP script that is going to be compiled and executed at server
; start-up.
; https://php.net/opcache.preload
;opcache.preload=

; Preloading code as root is not allowed for security reasons. This directive
; facilitates to let the preloading to be run as another user.
; https://php.net/opcache.preload_user
;opcache.preload_user=

; Prevents caching files that are less than this number of seconds old. It
; protects from caching of incompletely updated files. In case all file updates
; on your site are atomic, you may increase performance by setting it to "0".
;opcache.file_update_protection=2

; Absolute path used to store shared lockfiles (for *nix only).
;opcache.lockfile_path=/tmp
```

**Comment by ghost** [comment]
@nicolas-grekas 

After many hours of investigating, I finally found the origin of the issue.

It appears that the **Symfony adapter `PhpFilesAdapter`** does not work with the **XDebug extension 3.1.2 under PHP 8.1**. I even reinstalled all my PHP and XDebug versions to be sure.

On the other hand, there is **NO PROBLEM** between the **XDebug extension 3.1.2 under PHP 8.0** and the **Symfony adapter `PhpFilesAdapter`**.

Here are the results of my tests.
| PHP version          | XDebug extension version | XDebug extension enabled | Result    |
|----------------------|--------------------------|--------------------------|-----------|
| 8.0.13 (Thread Safe) | 3.1.2 (Thread Safe)      | NO                       | OK        |
| 8.0.13 (Thread Safe) | 3.1.2 (Thread Safe)      | YES                      | OK        |
| 8.1.0 (Thread Safe)  | 3.1.2 (Thread Safe)      | NO                       | OK        |
| 8.1.0 (Thread Safe)  | 3.1.2 (Thread Safe)      | YES                      | ISSUE !!! |
| 8.1.2 (Thread Safe)  | 3.1.2 (Thread Safe)      | NO                       | OK        |
| 8.1.2 (Thread Safe)  | 3.1.2 (Thread Safe)      | YES                      | ISSUE !!! |

Here are the screenshot of my tests.
![Test](https://user-images.githubusercontent.com/59090546/151793655-574275fb-4303-49f7-9dcb-fea2941532c6.png)

Does anyone have an idea how to solve the problem?

My OS is Windows 10.

[XDebug downloads](https://xdebug.org/download)
[Windows downloads](https://windows.php.net/download)

**Comment by nicolas-grekas** [comment]
Great findings!
I guess you should report that to https://github.com/xdebug/xdebug now.

**Comment by ghost** [comment]
> Great findings! I guess you should report that to https://github.com/xdebug/xdebug now.

Done.

[https://bugs.xdebug.org/view.php?id=2060](https://bugs.xdebug.org/view.php?id=2060)

**Comment by ghost** [comment]
Fixed in [XDebug 3.1.4](https://xdebug.org/updates#x_3_1_4).
